compiler/optimizations/gpuTransforms.cpp

```diff
@@ -2144,8 +2155,20 @@ static void cleanupPrimitives() {
       // uses of the primitive, which we process by removing the primitive but keeping
       // the copy.
       cleanupTaskIndependentCapturePrimitive(callExpr);
-    }
-    else if(callExpr->isPrimitive(PRIM_GPU_SET_BLOCKSIZE)) {
+    } else if (callExpr->isPrimitive(PRIM_GPU_SET_BLOCKSIZE) ||
+               callExpr->isPrimitive(PRIM_ASSERT_GPU_ELIGIBLE)) {
+      callExpr->remove();
+    } else if (callExpr->isPrimitive(PRIM_ASSERT_ON_GPU)) {
+      if (!usingGpuLocaleModel()) {
+        USR_FATAL_CONT(callExpr, "@assertOnGpu encountered in non-GPU compilation");
```

Comment:
I am still trying to decide whether this should be an error or warning. Technically, my `assertOnGpu`-tagged loop could be out of the dynamic execution path. That could appear in code like

```chpl
if here.gpus.size == 0 {
  foreach
}
else {
  @assertOnGpu
  foreach
}
```

I know that this is not the ideal way of performing that check, but it is perfectly valid nonetheless.

As a sidenote, Andy has this dream where the GPU locale model will be the only locale model, and you'll make decisions like the one above. We shouldn't make decisions based on that, but thought I should mention.

The bottom line to me is, if you have code like that today with the buggy `assertOnGpu`, it would work. I believe it only failed during execution time with the `flat` locale model, but I may be misremembering.

Reply:
To make it a runtime warning and allow the program to compile, I needed to add an implementation of `assert_on_gpu` for CPU specifically. I did so by tacking on to `chpl-gpu.h`, but I'm not sure that's right. I also moved the warning into module code and added a config param to disable it. Please let me know what you think.

Reply:
Looked at the current solution and I am ok with it. Thanks for making the change.